### PR TITLE
Spring boot shoping cart

### PR DESCRIPTION
**Pull request by Kulibchenko**

### CODE REVIEW

File: src/main/java/com/example/springboot/service/ShoppingCartService.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.example.springboot.service;
+
+import com.example.springboot.dto.CartItemDto;
+import com.example.springboot.dto.CreateCartItemRequestDto;
+import com.example.springboot.dto.ShoppingCartDto;
+import com.example.springboot.dto.UpdateCartItemRequestDto;
+import org.springframework.data.domain.Pageable;
+
+public interface ShoppingCartService {
+    CartItemDto save(CreateCartItemRequestDto requestDto, Long id);
```
Comment (Rommelua) [inline src/main/java/com/example/springboot/service/ShoppingCartService.java]:
```suggestion
    CartItemDto addToCart(CreateCartItemRequestDto requestDto, Long id);
```

File: src/main/java/com/example/springboot/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+@RestController
+@RequestMapping(value = "/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final UserService userService;
+
+    @Operation(summary = "Add items", description = "Add books to shoppingCart")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping
+    public CartItemDto createCartItem(@RequestBody @Valid CreateCartItemRequestDto requestDto,
```
Comment (Rommelua) [inline src/main/java/com/example/springboot/controller/ShoppingCartController.java]:
```suggestion
    public CartItemDto addToCart(@RequestBody @Valid CreateCartItemRequestDto requestDto,
```

File: src/main/java/com/example/springboot/service/ShoppingCartServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    public CartItemDto save(CreateCartItemRequestDto requestDto, Long id) {
+        Optional<ShoppingCart> shoppingCart = shoppingCartRepository.findByUserId(id);
+        if (shoppingCart.isPresent()) {
+            CartItem cartItem = cartItemMapper.toModel(requestDto);
+            if (bookRepository.findById(cartItem.getBook().getId()).isPresent()) {
+                cartItem.setShoppingCart(shoppingCart.get());
+                return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+            }
+        }
+        return null;
```
Comment (Rommelua) [inline src/main/java/com/example/springboot/service/ShoppingCartServiceImpl.java]: never return null

File: src/main/java/com/example/springboot/service/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        if (optionalShoppingCart.isPresent()) {
+            Set<CartItemDto> cartItemsDto = cartItemRepository.findAllByShoppingCartId(id).stream()
+                    .map(cartItemMapper::toDto)
+                    .collect(Collectors.toSet());
+            ShoppingCartDto shoppingCartDto = new ShoppingCartDto();
+            shoppingCartDto.setId(id);
+            shoppingCartDto.setCartItemsDto(cartItemsDto);
+            return shoppingCartDto;
+        }
+        return null;
```
Comment (Rommelua) [inline src/main/java/com/example/springboot/service/ShoppingCartServiceImpl.java]: never return null

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/springboot/dto/CartItemDto.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.example.springboot.dto;
+
+import lombok.Data;
+
+@Data
+public class CartItemDto {
+    private Long id;
+    private Long bookId;
+    private Long quantity;
+
```
Comment (koval-kn) [inline src/main/java/com/example/springboot/dto/CartItemDto.java]: redundant blank line

File: src/main/java/com/example/springboot/dto/CreateCartItemRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.example.springboot.dto;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CreateCartItemRequestDto {
+    @NotNull
```
Comment (koval-kn) [inline src/main/java/com/example/springboot/dto/CreateCartItemRequestDto.java:9]: it's better to ident annotated fields
```suggestion
    @NotNull
    private Long bookId;
    
    @NotNull
    @Min(value = 1)
    private Long quantity;
```

File: src/main/java/com/example/springboot/model/CartItem.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import org.hibernate.annotations.Where;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id =?")
+@Where(clause = "is_deleted = false")
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
```
Comment (koval-kn) [inline src/main/java/com/example/springboot/model/CartItem.java:23]: it's better to ident annotated fields
```suggestion
    @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @ManyToOne
    @JoinColumn(name = "shopping_cart_id", nullable = false)
    private ShoppingCart shoppingCart;
    
    @ManyToOne
    @JoinColumn(name = "book_id", nullable = false)
    private Book book;
    
    @Column(nullable = false)
    private Long quantity;
```

File: src/main/java/com/example/springboot/model/ShoppingCart.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
```
Comment (koval-kn) [inline src/main/java/com/example/springboot/model/ShoppingCart.java:19]: it's better to ident annotated fields
```suggestion
    @Id
    private Long id;
    
    @OneToOne
    @JoinColumn(name = "id", nullable = false)
    private User user;
    
    @OneToMany(mappedBy = "shoppingCart")
    private Set<CartItem> cartItems = new HashSet<>();
```

File: src/main/java/com/example/springboot/model/ShoppingCart.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    private Long id;
+    @OneToOne
+    @JoinColumn(name = "id", nullable = false)
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
+    private Set<CartItem> cartItems = new HashSet<>();
+
```
Comment (koval-kn) [inline src/main/java/com/example/springboot/model/ShoppingCart.java]: redundant blank line

File: src/main/java/com/example/springboot/repository/shoppingcart/ShoppingCartRepository.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package com.example.springboot.repository.shoppingcart;
+
+import com.example.springboot.model.ShoppingCart;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    Optional<ShoppingCart> findByUserId(Long id);
+
+    //Optional<ShoppingCart> findALl(Pageable pageable, @Param("id")Long id);
```
Comment (koval-kn) [inline src/main/java/com/example/springboot/repository/shoppingcart/ShoppingCartRepository.java]: YAGNI - you ain't gonna need it :)

**Review by koval-kn — CHANGES_REQUESTED** [review]

**Pull request merged**